### ISSUE TITLE
[FEATURE] Custom placeholder patterns in files

### ISSUE THREAD

**Issue by chrishumanitec** [score-spec/spec#108]
### Detailed description

When using the `files` feature in Score, it is often useful to embed placeholders. Unfortunately, the placeholder syntax in Score `${...}` is very commonly used and so can result in ambiguities.

It is possible to escape all of the other non-score placeholders, but this results in potentially major changes to a config file just to adopt score. A better approach would be to be able to define a custom pattern for a placeholder in a file - for example `%{...}` that could leave the existing file intact except for actual score placeholders.

Examples for config files that use `${...}` as a variable expansion pattern:
- [Spring Boot](https://www.baeldung.com/spring-boot-auto-property-expansion)
- [Shell Scripts](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html)

### Context

Adopting score should not require major changes to existing configuration - for example escaping existing variable expansions. This could for example mean that the configuration file does not work with existing local tooling.

This only makes sense for files rather than variables - as variables are only in the score of the score file anyway.

*NOTE* Score already supports `noExpand` [on the files object](https://docs.score.dev/docs/score-specification/score-spec-reference/#container-definition) to avoid processing placeholders.

### Possible implementation

There are a number of approaches that could be taken.

1. Allow changing of the leading `$` with another character. E.g `%` for `%{...}` style placeholders.
2. Supply start and end placeholders entirely. E.g. `<<` and `>>` for `<<...>>` style placeholders.

1 is simpler but 2 is more flexible as there could be unforseen situations where `{` is an integral part of the configuration language and a non overlapping prefix cannot be ruled out.

**Comment by astromechza** [comment]
Interesting. Prior art I've found are things like:

**Go text/template**

Go template language allows the delimiters to be set at the Parser level: https://pkg.go.dev/text/template#Template.Delims. This applies across all Go template files involved in simultaneous interpretation and cannot be turned on and off per template. Since Score allows team A to pull in a Score file published by team B, we would need to support this per Score file to avoid mandating the same delimiters.

Consul, a user of text/template, did a similar thing to expose this: https://github.com/hashicorp/consul-template/pull/615

** TODO more **

**Comment by astromechza** [comment]
Two things we could do while testing/figuring this out:

1. Generalize the placeholders lib in score-go to support custom delimeters. Would we still support escaping a placeholder?
2. Implement a score metadata annotation in score-compose that allows these delimiters to be set.

```
apiVersion: score.dev/v1b1
metadata:
  name: my-workload
  annotations:
    compose.score.dev/experiment-placeholder-delimiters: "%{ }"
```

**Comment by Abhishek9639** [comment]
Hi @mathieu-benoit and @chris-stephenson,
I’d like to work on this issue if it’s available. The problem and proposed approaches make sense to me, and I’m interested in implementing a flexible placeholder pattern solution. 
Please let me know if I can take this up.
Thanks

**Comment by chris-stephenson** [comment]
Hi @Abhishek9639 ,

You can pick this up, however be aware that it implies a proposal first for which implementation to use and how the spec schema should be updated to use it. Of you want to pick it up, you could create the proposal and then share it at the next score community meeting for example.

The proposal would need examples and justification for the approach. Ideally some examples of prior art Bryony go/template would be good too.

**Comment by Abhishek9639** [comment]
Thanks @chris-stephenson, 
sounds good. I'll put together a proposal with the implementation approach, schema changes, and some prior art examples. Will share it here before the next community meeting for feedback.

**Comment by Abhishek9639** [comment]
Hey @chris-stephenson and @mathieu-benoit and @astromechza,

Here's my initial proposal for custom placeholder delimiters. Happy to iterate on this based on your feedback.

### The problem

Score's `${...}` syntax for file placeholders conflicts with many config formats Spring Boot properties, shell scripts, Docker Compose env vars, etc. The existing workarounds (`$$` escaping or `noExpand: true`) either require modifying every config file or disable Score expansion entirely, which doesn't work when you need both Score placeholders and literal `${...}` in the same file.

### Proposed approach: custom start/end delimiters per file

I'm going with **Approach 2** from the issue let users define a custom pair of start/end delimiters on each file entry. This way, Score looks for e.g. `%{resources.db.host}` instead of `${resources.db.host}`, and leaves any `${...}` alone.

I considered two options for where to put this:

**Option A - Schema-level property on `containerFile`:**
```yaml
containers:
  app:
    image: myapp
    files:
      "/app/config.properties":
        source: ./config.properties
        expandDelimiters:
          start: "%{"
          end: "}"
```

**Option B - Metadata annotation (as @astromechza suggested):**
```yaml
apiVersion: score.dev/v1b1
metadata:
  name: my-workload
  annotations:
    compose.score.dev/experiment-placeholder-delimiters: "%{ }"
```

I'm leaning towards **Option A** because:
- It's per-file, so different files in the same workload can use different delimiters
- It's part of the spec, not implementation-specific works across score-compose, score-k8s, etc.
- It sits next to `noExpand` which is already on `containerFile`, so it's consistent

That said, Option B is simpler and doesn't require a spec change. Could be useful as an experimental first step before committing to a schema change. Would love to hear your thoughts on this tradeoff.

### Real-world examples

**Spring Boot** - shell `${...}` and Score placeholders in the same file:
```yaml
files:
  "/app/application.properties":
    source: ./application.properties
    expandDelimiters:
      start: "<%{"
      end: "}%>"
```
```properties
# These stay as-is (Spring Boot resolves them)
spring.datasource.url=jdbc:postgresql://${DB_HOST}:${DB_PORT}/${DB_NAME}

# These get expanded by Score
app.db.host=<%{resources.db.host}%>
app.db.port=<%{resources.db.port}%>
```

**Shell script:**
```yaml
files:
  "/entrypoint.sh":
    source: ./entrypoint.sh
    expandDelimiters:
      start: "<<"
      end: ">>"
```
```bash
#!/bin/bash
echo "Running as ${USER}"  # shell variable, left alone
export DB_HOST=<<resources.db.host>>  # Score placeholder, expanded
```

### Prior art

- **Go `text/template`** - `template.Delims("[[", "]]")` per template instance ([docs](https://pkg.go.dev/text/template#Template.Delims))
- **Consul Template** - `left_delimiter` / `right_delimiter` per template block ([hashicorp/consul-template#615](https://github.com/hashicorp/consul-template/issues/615))
- **Jinja2** - `variable_start_string` / `variable_end_string` on the Environment ([docs](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Environment))
- **Ruby ERB** - fixed `<%= %>` delimiters, no customization (users switch to Liquid or Mustache when they need different syntax)
- **Helm** - no custom delimiter support, uses `$$` escaping only. This is a known pain point with lots of community workarounds.

The pattern across Go template, Consul, and Jinja2 is the same: a configurable start/end pair, set per template or environment. That's what I'm proposing here.

### Interaction with existing features

- `noExpand: true` takes priority if set, custom delimiters are ignored
- `binaryContent` doesn't support expansion already same behavior
- Default (no `expandDelimiters`) works exactly as today, fully backward compatible

### On escaping (@astromechza's question)

If custom delimiters are set, the default `$$` escaping for `${...}` wouldn't be needed anymore since `${...}` is no longer special. But we should still support escaping the custom delimiters e.g. if `start` is `%{`, then `%%{` could produce a literal `%{`. Need to think about whether this adds too much complexity though.

### Open questions

1. **Option A vs B** - spec property or annotation? Or start with annotation, promote to spec later?
2. **Workload-level default** - should we support setting delimiters at the workload level too (to avoid repeating per file)?
3. **Naming** - `expandDelimiters`, `placeholderSyntax`, `substitutionDelimiters`? Open to suggestions.

Looking forward to your feedback. Happy to present this at the next community meeting too.
Thanks

**Comment by mathieu-benoit** [comment]
Thanks, @Abhishek9639! Just added it to our agenda for our next community meeting: https://github.com/score-spec/spec/discussions/185.

**Comment by Abhishek9639** [comment]
Hey @chris-stephenson and @mathieu-benoit,
Following up on starting with Option B. I went through the code and wanted to run a plan past you before opening anything.

## Why Option B looks like the right starting point
No spec or schema change needed. It fits the existing `compose.score.dev/experiment-*` annotation pattern already used for `publish-port,` `workload-name`, and the `route-provisioner-*` family. We can ship it, get real usage at the community meetings, and promote to Option A once we actually know what delimiter choices survive contact with users. It also lets the underlying score-go change (parameterizable delimiters) land in isolation, without spec churn.

## Implementation plan

`score-go` (framework/substitution.go):

- The regex today is hardcoded: `regexp.MustCompile(`$((?:$?{([^}]*)})|$)`)`.
- Add `NewSubstituterWithDelimiters(start, end string)` that builds the regex from `regexp.QuoteMeta(start) + "(.*?)" + regexp.QuoteMeta(end)` and stores it on the `Substituter`.
- The zero-value `Substituter{}` keeps the current regex and behavior, so nothing existing breaks.
- Tests: a matrix of `(start, end)` pairs (`${ }, %{ }, << >>, <%{ }%>`), checking that literal `${...}` is left alone when custom delimiters are in use, and that the default path is byte-identical to current output.

`score-compose` (internal/compose/convert.go):

- Read the annotation from workload metadata once, build a `Substituter` with those delimiters, use it for file content expansion. Annotation absent, nothing changes.
- One thing on parsing I want to flag. The original sketch was `"%{ }"`, single annotation, space-separated. That breaks the moment someone wants a delimiter with a space in it. Two annotations seems safer:

> compose.score.dev/experiment-placeholder-start
> compose.score.dev/experiment-placeholder-end

Required together, validation error if only one is set. Happy to keep the single-annotation form if you'd rather, but wanted to surface it.
- `noExpand: true` still wins. `binaryContent` is unaffected.
- Tests in the existing `convert_test.go `style: annotation set, annotation unset, annotation +` noExpand`, malformed annotation.

## On escaping

With custom delimiters set, `${...}` is just text, so the `$$` unescaper doesn't apply to that case. My instinct is to skip escaping the custom delimiters themselves in v1. Go's `text/template` doesn't escape custom delims either, and a meta-escape rule is the kind of thing that's hard to walk back once it's shipped. Easy to add later if someone actually hits it.

## What this leaves for Option A

Two things Option B can't do: per-file delimiters, and working in score-k8s (which would need its own annotation). Those are basically the case for promoting to a spec property later, and by then we'll have real data on whether anyone needs them.

## One thing to decide before I start

Single annotation (`"%{ }"`) or two (`-start` / `-end`)? Two is my preference, but your call.

If the plan looks reasonable I'll open the `score-go` PR first (parameterized substituter, default unchanged), then a `score-compose` PR that consumes it. Can also walk through this at the next community meeting.

Thanks!

**Comment by chris-stephenson** [comment]
Thanks @Abhishek9639 for documenting this out.

- I think this makes sense as an experiment.
- I agree with 2 annotations rather than one that needs parsing.

Your argument for go templates not having escpaing does not quite add up. This is because escaping in Go templates is itself a template. This is not the case with placeholders as we're not really doing templating.

For example, if I want to output the string `{{ this should not be processed }}` in go templates, I would do this:
```
{{"{{"}} this should not be processed {{"}}"}}
```
So you can escape go templates - but by using a feature of templating.

**Comment by Abhishek9639** [comment]
Thanks @chris-stephenson, and fair point on the Go templates comparison. You're right, my analogy was off. Go templates can emit a literal `{{` because the template engine itself parses the file, so `{{"{{"}}` is a template expression that prints the string `{{`. Score placeholders don't have that route since we're just running a regex over raw text, so "Go doesn't escape" was the wrong way to defend skipping it.

So escaping is a real open call. Three options I can see:

1. Doubled-prefix escape. If `start` is `%{`, then `%%{` produces a literal `%{`. Cheap to implement (post-process matches), and it's symmetric with how `$$` → `$` already works in Score today, so it's not a new concept for users to learn.
2. No escape, just document the constraint that delimiters can't appear literally in your file. Simplest, but probably the first thing someone complains about.
3. Backslash escape (`\%{` → `%{)`. Familiar from shell, but it introduces a new special character that wasn't there before.

I'd go with option 1. It's closest to what users already see in Score, and if we change our minds later it's easy to drop.

To summarise what's settled now:

- Two annotations (`-start` / `-end`), required together.
- Experiment-prefixed, score-compose only for v1.
- `${...}` stops being special when custom delimiters are set; the existing `$$` escape stops applying in that mode.
- Doubled-prefix escape for the custom delimiters, pending your sign-off on option 1.

If option 1 works for you, I'll start on the `score-go` PR: parameterized substituter, default behavior unchanged, plus the escape handling.

Thanks again for the correction!

**Comment by chris-stephenson** [comment]
I would just be a bit careful here - you've jumpped to a solution that does not work with your examples.  How can you infer an escape for `<< >>` pair?

**Comment by Abhishek9639** [comment]
Hey @chris-stephenson,
You're right, that doesn't hold up. The `$$` → `$` rule worked because the start delimiter began with a single character, so doubling that character was unambiguous. With `<< >>` there's no clean "doubled prefix": `<<<` could be read as `<` then `<<`, or as escape-then-start. Asymmetric delimiters like `<%{ }%>` are even worse. The analogy didn't survive contact with my own examples.

For an experiment, I now lean toward no escape at all in v1. The whole point of custom delimiters is that the user picks something that doesn't collide with their file content. If someone picks `<<` and their file genuinely needs a literal `<<`, the right move is to pick different delimiters (e.g. `[[ ]]`), not learn a Score-specific escape rule. We document the constraint clearly: delimiters must not appear literally in the file, choose ones that don't collide.

If escape does become a real pain point once people use this, the most generalisable mechanism I can think of is a backslash prefix on the start delimiter (`\<<` → `<<`, `\%{` → `%{`, `\<%{` → `<%{`). It works regardless of delimiter shape. But that's a v2 decision based on actual feedback, not something I'd bake in now.

Revised plan:

- Two annotations, required together.
- No escape in v1, with a docs note on choosing non-colliding delimiters.
- `${...}` stops being special when custom delimiters are set, and the existing `$$` escape stops applying in that mode.
- Escape design deferred until we have a real complaint.

If that's reasonable, I'll start on the score-go PR. Thanks for catching this before I started writing the regex.

Thanks!
